numpy/f2py/src/fortranobject.c

```diff
@@ -729,44 +836,41 @@ PyArrayObject* array_from_pyobj(const int type_num,
                 sprintf(mess+strlen(mess),"%" NPY_INTP_FMT ",",dims[i]);
             strcat(mess, ")");
             PyErr_SetString(PyExc_ValueError,mess);
+            Py_DECREF(descr);
             return NULL;
         }
-        arr = (PyArrayObject *)
-            PyArray_New(&PyArray_Type, rank, dims, type_num,
-                        NULL,NULL,1,
-                        !(intent&F2PY_INTENT_C),
-                        NULL);
-        if (arr==NULL) return NULL;
+        arr = (PyArrayObject *)                                      \
+          PyArray_NewFromDescr(&PyArray_Type, descr, rank, dims,
+                               NULL, NULL, !(intent & F2PY_INTENT_C), NULL);
+        if (arr == NULL) {
+          Py_DECREF(descr);
+          return NULL;
+        }
+        if (PyArray_ITEMSIZE(arr) != elsize) {
+          strcpy(mess, "failed to create intent(cache|hide)|optional array");
+          sprintf(mess+strlen(mess)," -- expected elsize=%d got %" NPY_INTP_FMT, elsize, (npy_intp)PyArray_ITEMSIZE(arr));
+          PyErr_SetString(PyExc_ValueError,mess);
+          Py_DECREF(arr);
+          return NULL;
+        }
         if (!(intent & F2PY_INTENT_CACHE))
-            PyArray_FILLWBYTE(arr, 0);
+          PyArray_FILLWBYTE(arr, 0);  // TODO: should we fill fortran character arrays with blanks?
```

Comment:
I think a blank default would make sense.

Reply:
This is actually tricky: (i) we don't know if the consumer of the hidden array is actually a Fortran program. (ii) this change may have BC implications.

If filling with blanks is indeed required for a Fortran program, this can be done explicitly from the signature file using the initialization expression. So, leaving here as it is.